### PR TITLE
feat: V2 — sidebar sync, agent hierarchy, quality tracking

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

Three V2 features for the cmux MCP agent lifecycle layer:

- **T17 Sidebar sync:** 3-channel push (set-status, set-progress, log). In-memory diff — only pushes changed entries. Lifecycle event logging (spawned/done/errored).
- **T18 Agent hierarchy:** parent_agent_id + spawn_depth on AgentRecord. Cascade kill (DFS post-order) for intentional kills, orphan survival for crash (children keep running independently). MAX_SPAWN_DEPTH=2, MAX_CHILDREN=10.
- **T19 Quality tracking:** quality field (unknown/verified/suspect/degraded). Parses context % from agent screen. At 80%: /compact for depth-0, kill+log for depth>0. Optional maxCostPerAgent config (stored, not yet enforced).

## Design

Based on converged 22-decision design doc with 3 research rounds (MCP discriminated unions, agent hierarchy patterns, cmux sidebar Swift source reading).

## Test plan

- [x] 210 tests total, all passing (32 new)
- [x] Sidebar sync: 8 tests (diff detection, progress, logging)
- [x] Hierarchy: 11 tests (parent tracking, cascade kill, orphan survival, depth/breadth guards)
- [x] Quality: 13 tests (context parsing, degradation, circuit breaker, kill-only behavior)
- [x] TypeScript strict, no any

## Summary by CodeRabbit

* **New Features**
  * Agent hierarchy: parent/child links, spawn limits, subtree retrieval, and cascade kill for entire subtrees
  * Sidebar sync: real-time status, progress, and one-time lifecycle event logging
  * Quality tracking: context-based quality states and automated actions; per-agent cost/quality metadata
  * Public sweep operation and client logging API; ability to update agent records

* **Bug Fixes**
  * Safer command handling by sanitizing repository names and paths

* **Tests**
  * Comprehensive suites for hierarchy, sidebar sync, quality tracking, and related behaviors

### CODE REVIEW

**Review by EtanHey — COMMENTED** [review]
## Pass 1/3: Design Alignment — 22 Converged Decisions

Reviewing against `docs.local/plans/phase5-v2-synthesis.md`. For future AI reviewers: this design doc has 3 layers — the original synthesis, a User Refinement (explicit actions), and a V2 Ideation Round (agent graph, sidebar sync, discriminated unions). The V2 Ideation Round is the authoritative source.

---

### Q1: Does sidebar sync use `sidebar-state` for diffing?

**DEVIATION.** The design doc (Section 3, "Level-Triggered Reconciliation") describes reading sidebar-state to diff. The implementation uses an **in-memory `sidebarSnapshot` Map** instead. Each agent's status string is cached; only changed entries are pushed.

This is arguably **better** than the spec. Reading `sidebar-state` every 5s adds a cmux CLI call per sweep. The in-memory diff is O(n) with zero I/O. The only downside: if the MCP server restarts, it re-pushes ALL agent statuses on the first sweep (cache is empty). Since `set_status` is idempotent, this is a non-issue — exactly what the design doc's "level-triggered" model says should happen.

**Verdict: Acceptable deviation. Document the rationale in a code comment.**

---

### Q2: Does hierarchy use child→parent reference?

**PARTIAL.** The design doc (Section 2, "Agent Graph") proposes BOTH:
- `parent_agent_id` on AgentRecord (child→parent, K8s-style)
- `private children = new Map<string, Set<string>>()` on AgentRegistry (parent→children index)

The implementation has `parent_agent_id` (child→parent) but **no parent→children index**. Instead, `getChildren()` filters the entire agent map:

```typescript
getChildren(parentId: string): AgentRecord[] {
  return [...this.agents.values()].filter(
    (a) => a.parent_agent_id === parentId,
  );
}
```

This is O(n) per lookup. For `cascadeKill` on a tree of depth d with branching factor b, it's O(n × d × b). With MAX_CHILDREN=10 and MAX_SPAWN_DEPTH=2, the worst case is ~30 agents × 3 levels = ~90 iterations. Fine for now, but the design explicitly called for the index.

**Verdict: Add a `// TODO: add parent→children index when n > 50 matters` comment. Ship as-is.**

---

### Q3: Does quality tracking parse context % correctly?

**YES, after a regex fix applied in the same PR.** The original regex had a greedy `\S*` that consumed digits. Fixed to:

```typescript
text.match(/(\d{1,3})%\s*context/i) ?? text.match(/context[^%]*?(\d{1,3})%/i)
```

Covers: `"80% context"`, `"context 75%"`, `"80% context remaining"`, `"Claude Code ● 85% context Model: sonnet"`. Rejects: values > 100, non-context percentages.

**One concern:** The actual Claude Code status bar format hasn't been empirically verified. If CC changes its format (e.g., `"context window: 80%"` or uses Unicode characters), the regex silently fails (returns null → no action). This is fail-safe (no false degradation), but the agent could run to 100% context without intervention.

**Verdict: Correct. Add an integration test with a real CC screen capture when available.**

---

### Q4: Are cascade vs orphan policies correct?

**CASCADE: CORRECT.** `cascadeKill()` calls `getSubtree()` which returns DFS post-order (children before root), then iterates with `stopAgent()`. Children die first. Best-effort: if one child stop fails, continues to next. Matches the design doc's `killBranch` spec.

**ORPHAN-ON-CRASH: PARTIALLY CORRECT.** When a parent surface disappears, `reconcileSurfaces()` transitions ONLY that agent to error. Children keep running because they have their own surfaces. This is correct.

However, **orphaned children retain a `parent_agent_id` pointing to a dead agent.** The PR description claims "orphan reparenting for crash_detected" but no reparenting is implemented. If you call `getSubtree(dead_parent_id)`, it still finds the orphans (the dead parent is still in the registry in "error" state). If the dead parent's state file is cleaned up, `getSubtree` silently drops the orphans from the tree.

**This is not a bug but it IS misleading.** The PR description should say "orphan survival" not "orphan reparenting."

**Verdict: Update PR description. Add a `// TODO: reparent orphans to root or null` comment in reconcileSurfaces.**

---

### Q5: PR description accuracy

The PR description has three claims not matched by the implementation:

| Claim | Reality |
|-------|---------|
| "orphan reparenting for crash_detected" | Orphan survival only — no reparenting |
| "kill+respawn for depth>0" at 80% context | Kill only — no respawn |
| "MAX_SPAWN_DEPTH=2, MAX_CHILDREN=10 (configurable)" | They are `const` — not configurable at runtime |

These aren't bugs but the description should be accurate for future reviewers.

**Review by EtanHey — COMMENTED** [review]
## Pass 2/3: Security, Correctness, Edge Cases

---

### ISSUE 1: Command injection in `buildLaunchCommand` [SECURITY]

```typescript
function buildLaunchCommand(cli: CliType, repo: string): string {
  const cdCmd = `cd ~/Gits/${repo}`;
  // ...
}
```

`repo` is interpolated into a shell command without sanitization. If `repo` contains shell metacharacters (semicolons, backticks, `$()` subshells), they execute in the user's shell.

**Attack vector:** An LLM calling `spawnAgent({ repo: "brainlayer; malicious-command", ... })` could execute arbitrary commands. Since the MCP server accepts repo names from LLM tool calls, this is a real attack surface.

**Fix:**

```typescript
function buildLaunchCommand(cli: CliType, repo: string): string {
  if (!/^[a-zA-Z0-9._-]+$/.test(repo)) {
    throw new Error(`Invalid repo name: ${repo}`);
  }
  const cdCmd = `cd ~/Gits/${repo}`;
  // ...
}
```

Or use `shellEscape()`. The allowlist regex is safer — repo names are always alphanumeric + hyphens.

**Severity: HIGH.** Fix before merge.

---

### ISSUE 2: `generateAgentId` collision [BUG]

```typescript
export function generateAgentId(model: string, repo: string): string {
  const ts = Math.floor(Date.now() / 1000);
  const slug = repo.replace(/[^a-zA-Z0-9-]/g, "-");
  return `${model}-${slug}-${ts}`;
}
```

Two agents spawned in the same second with the same model+repo get **identical IDs**. The second agent silently overwrites the first in both the state file and registry. This was already hit in testing — the `MAX_CHILDREN` test had to use unique repo names per child as a workaround.

**Fix:** Add a random suffix:

```typescript
const rand = Math.random().toString(36).slice(2, 6);
return `${model}-${slug}-${ts}-${rand}`;
```

Or use a monotonic counter on the AgentEngine instance.

**Severity: MEDIUM.** Not a security issue but causes silent data loss. The test workaround masks the production bug.

---

### ISSUE 3: `getSubtree` has no cycle detection [EDGE CASE]

```typescript
getSubtree(rootId: string): AgentRecord[] {
  const collect = (id: string) => {
    const children = this.getChildren(id);
    for (const child of children) {
      collect(child.agent_id);
    }
    // ...
  };
  collect(rootId);
}
```

If a corrupt state file creates a circular parent chain (A is parent of B, B is parent of A), this infinite loops and crashes the MCP server. Unlikely with normal operation, but possible if state files are hand-edited or corrupted.

**Fix:** Add a visited `Set<string>`:

```typescript
getSubtree(rootId: string): AgentRecord[] {
  const result: AgentRecord[] = [];
  const visited = new Set<string>();
  const collect = (id: string) => {
    if (visited.has(id)) return;
    visited.add(id);
    // ...
  };
```

**Severity: LOW.** Defensive programming. Easy fix.

---

### ISSUE 4: `readScreen` called for EVERY non-terminal agent on EVERY sweep [PERFORMANCE]

In `syncSidebar()`, quality tracking calls `readScreen` for every non-terminal agent:

```typescript
if (!TERMINAL_STATES.has(state)) {
  try {
    const screen = await this.client.readScreen(surface_id, { lines: 5 });
```

With 10 agents and 5s sweeps, that's 120 `readScreen` calls/minute. The sidebar research doc confirmed 120 `set_status` calls/min is fine, but `readScreen` is heavier — it captures terminal content.

**Recommendation:** Only check context % every Nth sweep (e.g., every 6th = every 30s) or only for agents in "working" state. A simple counter:

```typescript
private sweepCount = 0;
// In syncSidebar:
const checkQuality = ++this.sweepCount % 6 === 0;
```

**Severity: LOW.** Not blocking, but will matter at scale.

---

### ISSUE 5: Dead code — `deletion_intent` and `max_cost_per_agent` [CODE QUALITY]

- `deletion_intent: boolean` — Added to AgentRecord, stored on disk, has a dedicated `updateRecord` path. **Never set to true or checked anywhere.**
- `max_cost_per_agent: number | null` — Stored in record at spawn time. **Never read or enforced.**

Both are future hooks. They add fields to every state file and every test's `makeRecord` helper. Dead code in a PR is a review smell — it makes the diff larger without delivering value.

**Recommendation:** Either implement the enforcement logic now (even minimal) or remove the fields and add them when the enforcement is built. If keeping them, add `// TODO:` comments explaining what will use them.

**Severity: LOW.** Not harmful, but inflates the PR.

---

### ISSUE 6: `sidebarSnapshot` and `loggedEvents` grow unbounded [MEMORY]

These in-memory maps/sets are never pruned. After hundreds of agent spawn/done cycles, they accumulate entries for every agent that ever existed.

```typescript
private sidebarSnapshot = new Map<string, string>();
private loggedEvents = new Set<string>();
```

**Fix:** When an agent enters a terminal state AND the status has been pushed, remove it from `sidebarSnapshot` after a delay (e.g., keep for 60s for final status display, then prune). Or prune on every Nth sweep.

**Severity: LOW.** Only matters for long-running sessions with many agents.

**Review by EtanHey — COMMENTED** [review]
## Pass 3/3: Test Coverage Gaps & Summary

---

### Test coverage: What's tested well

- **Sidebar diff logic**: 8 tests covering push-on-change, no-push-on-same, lifecycle logs (spawned/done/error), progress bar ratio, double-log prevention. Solid.
- **Hierarchy guards**: Depth limit, breadth limit, parent-not-found. All the validation paths are tested.
- **getSubtree DFS ordering**: 3 tests (single node, parent+child, full tree) verify post-order traversal. Good.
- **cascadeKill best-effort**: Tests that failure of one child doesn't abort the rest.
- **Quality parsing**: 6 tests on `parseContextPercent` including edge cases (no match, >100, mixed text). Good unit coverage.
- **Quality actions**: Verifies `/compact` for depth-0, kill for depth-1, degraded quality update, no-action below 80%.

### Test coverage: What's missing

| Gap | Risk | Suggested test |
|-----|------|---------------|
| `buildLaunchCommand` with special chars in repo | Security (Issue 1) | `expect(() => buildLaunchCommand("claude", "a;b")).toThrow()` |
| `generateAgentId` collision | Data loss (Issue 2) | Spawn 2 agents same model+repo, verify different IDs |
| `getSubtree` with circular parent chain | Crash (Issue 3) | Write records A→B→A, verify getSubtree terminates |
| `readScreen` failure in quality tracking | Resilience | Mock readScreen to throw, verify sweep completes without error |
| `syncSidebar` with 0 agents | Edge case | Empty registry, verify no setProgress/setStatus called |
| Concurrent `runSweep` calls | Race condition | Two overlapping sweeps, verify no duplicate status pushes |
| `updateRecord` on nonexistent agent | Error path | `expect(() => stateMgr.updateRecord("ghost", {...})).toThrow()` |

The `readScreen` failure test is important — the code has a `catch {}` block for it, but it's untested. Silent catch blocks are exactly the kind of thing that should have test proof.

---

### Summary for future AI reviewers

**Architecture context:** This PR adds 3 features to the cmux MCP agent lifecycle layer. The cmux MCP is a TypeScript server that manages AI CLI agents (Claude Code, Codex, Gemini, etc.) running in cmux terminal panes. It tracks agent state via flat JSON files, maintains an in-memory registry, and projects state to cmux's sidebar UI.

**Key design decisions to know:**
1. State files (`~/.local/state/cmux-agents/{id}/state.json`) are the source of truth. Registry is a computed cache. Sidebar is a display projection.
2. Reconciliation sweep (every 5s) is the heartbeat — syncs disk→registry→sidebar.
3. Level-triggered > edge-triggered: the sweep catches any missed transitions.
4. Agents have a tree structure (parent→children via `parent_agent_id`). Intentional kill cascades down the tree (DFS post-order). Crash does NOT cascade — children survive (orphan-on-crash).
5. Quality tracking reads terminal content to detect context window saturation. Root agents get `/compact`, non-root agents get killed.

**What to watch for in future PRs:**
- `generateAgentId` collision is a known unfixed bug. Any test that spawns multiple agents with same model+repo in <1s will hit it.
- `deletion_intent` and `max_cost_per_agent` are scaffolding fields with no enforcement. They'll be wired up later.
- The `buildLaunchCommand` shell injection needs fixing before any user-facing deployment.
- The `readScreen` call frequency in quality tracking will need throttling for >10 agent scenarios.

---

### Verdict

**Do NOT approve yet.** Two items need fixing before merge:

1. **ISSUE 1 (command injection):** Add repo name validation in `buildLaunchCommand`. This is a real security surface since LLMs provide the repo name.
2. **ISSUE 2 (ID collision):** Add entropy to `generateAgentId`. The current workaround (unique repos in tests) masks a production bug.

Everything else is either acceptable deviation from the design doc or low-priority improvements that can be follow-up PRs.

**Test count:** 208 total (30 new), all passing. Good coverage for the happy paths. Missing coverage on error paths and edge cases listed above.

**Comment by EtanHey** [comment]
@greptileai review

**Comment by EtanHey** [comment]
@coderabbitai review

File: src/agent-engine.ts [hunk 1]
```diff
@@ -73,13 +100,119 @@ export class AgentEngine {
...
+            const updated = this.stateMgr.updateRecord(agentId, {
+              quality: "degraded",
+            });
+            this.registry.set(agentId, updated);
+
+            if (agent.spawn_depth === 0) {
+              // Root agent: send /compact
+              await this.client.send(surface_id, "/compact", {});
+              await this.client.sendKey(surface_id, "return", {});
+            } else {
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:178]: Both `send()` and `sendKey()` are `await`ed sequentially. The cmux socket processes commands in order — text delivery completes before the return key is sent. No race condition exists here. **Skipping.**

File: src/agent-engine.ts [hunk 2]
```diff
@@ -73,13 +100,119 @@ export class AgentEngine {
...
+        }
+      }
+    }
+
+    // Progress bar
+    if (total > 0) {
+      await this.client.setProgress(done / total, {
+        label: `agents ${done}/${total}`,
+      });
+    }
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:201]: 0% is semantically correct when zero agents are done. The label `"agents 0/1"` already communicates "in progress." A fake minimum (0.05) would be misleading — the progress bar should reflect reality. **Skipping.**

File: src/agent-registry.ts [hunk 3]
```diff
@@ -105,4 +109,31 @@ export class AgentRegistry {
...
+      const children = this.getChildren(id);
+      for (const child of children) {
+        collect(child.agent_id);
+      }
+      const agent = this.agents.get(id);
+      if (agent) result.push(agent);
+    };
+    collect(rootId);
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-registry.ts:154]: Valid. MAX_SPAWN_DEPTH=2 prevents real cycles but corrupted state files could still cause infinite recursion. Adding a `visited` set — cheap defensive fix.

File: src/agent-types.ts [hunk 4]
```diff
@@ -80,11 +92,25 @@ export function assertValidTransition(from: AgentState, to: AgentState): void {
...
+ * Matches patterns like "80% context", "context 80%", "80% context remaining"
+ */
+export function parseContextPercent(text: string): number | null {
+  const m =
+    text.match(/(\d{1,3})%\s*context/i) ??
+    text.match(/context[^%]*?(\d{1,3})%/i);
+  if (!m) return null;
+  const n = parseInt(m[1], 10);
+  return isNaN(n) || n < 0 || n > 100 ? null : n;
+}
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-types.ts:106]: Good catch. `\b` word boundary prevents "1234% context" from extracting "123" as a partial match. The `> 100` validation would reject it anyway, but the regex should not extract false matches in the first place. Fixing.

File: src/state-manager.ts [hunk 5]
```diff
@@ -113,6 +113,37 @@ export class StateManager {
...
+    };
+
+    const agentDir = join(this.baseDir, agentId);
+    const stateFile = join(agentDir, "state.json");
+    const tmpFile = join(agentDir, "state.json.tmp");
+    writeFileSync(tmpFile, JSON.stringify(updated, null, 2), "utf-8");
+    renameSync(tmpFile, stateFile);
+
+    return updated;
+  }
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/state-manager.ts:159]: Valid. Using `source: "updateRecord"` to distinguish from state-machine transitions. Quality degradation events should be traceable in the event log for incident debugging. Fixing.

File: tests/sidebar-sync.test.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    updated_at: "2026-03-14T03:40:00Z",
+    error: null,
+    parent_agent_id: null,
+    spawn_depth: 0,
+    deletion_intent: false,
+    quality: "unknown",
+    max_cost_per_agent: null,
+    ...overrides,
+  };
+}
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/sidebar-sync.test.ts:77]: Self-contained test files are a deliberate choice — when a test fails, all context is in one file. The `listStatus` drift is cosmetic (not called by sidebar sync code). A shared fixture would couple test suites together. **Skipping.**

File: tests/agent-hierarchy.test.ts [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    await expect(
+      engine.spawnAgent({
+        repo: "brainlayer",
+        model: "sonnet",
+        cli: "claude",
+        prompt: "Orphan",
+        parent_agent_id: "nonexistent-parent",
+      }),
+    ).rejects.toThrow("Parent agent not found: nonexistent-parent");
+  });
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/agent-hierarchy.test.ts:222]: **Pushing back.** The design doc (`phase5-v2-synthesis.md` Section 2) specifies **orphan survival**, not orphan reparenting:

> "Orphan-on-crash: when a parent surface disappears, only THAT agent transitions to error. Children are not affected because each child has its own separate surface."

The PR description incorrectly says "orphan reparenting" — will fix to say "orphan survival." No reparenting logic exists because the design deliberately leaves orphaned children running independently.

**Adding a test** that verifies children continue running when parent crashes (orphan survival).

File: tests/quality-tracking.test.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    await engine.runSweep();
+
+    // Should have been killed (sendKey c-c for graceful stop)
+    expect(mockClient.sendKey).toHaveBeenCalledWith("s:2", "c-c", {});
+    // Should log the context-limit kill
+    expect(mockClient.log).toHaveBeenCalledWith(
+      expect.stringContaining("context-limit"),
+      expect.objectContaining({ level: "warning" }),
+    );
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/quality-tracking.test.ts:200]: **Pushing back.** The implementation intentionally kills without respawning. The design doc quality tracking section specifies: "depth 0: send /compact, depth 1+: kill and log." Respawn is not in the design scope for v1.

The PR description incorrectly claims "kill+respawn" — will fix to say "kill+log." Adding a test that documents the kill-only behavior with a TODO for future respawn.

File: tests/sidebar-sync.test.ts [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+      (c) => typeof c[0] === "string" && c[0].startsWith("spawned:"),
+    );
+    expect(spawnedCalls).toHaveLength(1);
+    // "done" should appear exactly once
+    const doneCalls = logCalls.filter(
+      (c) => typeof c[0] === "string" && c[0].startsWith("done:"),
+    );
+    expect(doneCalls).toHaveLength(1);
+  });
+});
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/sidebar-sync.test.ts:279]: **Pushing back.** The implementation uses 3 channels (set-status, set-progress, log). `rename-workspace` and `report_meta_block` are **not implemented** in this PR — the sidebar research doc (`phase5-v2-cmux-sidebar-research.md`) identifies them as secondary channels for future enhancement.

The PR description incorrectly claims "5-channel push" — will fix to say "3-channel push (set-status, set-progress, log)." There is nothing to test for channels that do not exist in the code.

These are valid future enhancements but not missing coverage for the current implementation.

File: src/agent-engine.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:178]: @coderabbitai This is intentional. In cmuxlayer, `send()` and `sendKey()` are both awaited sequentially via the cmux Unix socket. The socket processes commands in FIFO order — text delivery completes before the return key is dispatched. The 5s reconciliation sweep catches any edge cases. This is a known tradeoff documented in the design. No delay needed between cmux send/sendKey pairs.

File: src/agent-engine.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/agent-engine.ts:201]: @coderabbitai In cmuxlayer, progress bars reflect exact completion ratios. 0% with label "agents 0/1" is correct — it means work started but nothing finished. We never use fake minimums. The progress bar is a data display, not a UX widget. The label carries the context.

File: tests/sidebar-sync.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/sidebar-sync.test.ts:77]: @coderabbitai In cmuxlayer, each test file is self-contained by design. When a test fails, all context (mock setup, helpers, assertions) is in one file. Shared fixtures would couple test suites and make failures harder to diagnose. The minor drift in mock fields (e.g., listStatus) is acceptable — it only matters when a test explicitly calls that method.

File: tests/agent-hierarchy.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/agent-hierarchy.test.ts:222]: @coderabbitai Important learning for this repo: cmuxlayer uses **orphan survival**, not orphan reparenting. When a parent agent crashes, children continue running independently because each agent has its own cmux surface. See `phase5-v2-synthesis.md` Section 2. This is the Kubernetes orphan GC pattern — ownerReferences exist but deletion does NOT cascade on crash. Cascade kill is only for intentional `cascadeKill()` calls. Added a test verifying this in b2b4da0.

File: tests/quality-tracking.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/quality-tracking.test.ts:200]: @coderabbitai Important learning for this repo: cmuxlayer quality tracking uses **kill-only** for depth>0 agents at the context limit. No auto-respawn. The design says "depth 1+: kill and log." Respawn requires agent session resume infrastructure (BrainLayer → JSONL → cli_session_id mapping) which is a future v2 feature. Added a test documenting this intentional behavior in b2b4da0.

File: tests/sidebar-sync.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/sidebar-sync.test.ts:279]: @coderabbitai Important learning for this repo: cmuxlayer sidebar sync currently uses 3 channels (set-status, set-progress, log). rename-workspace and report_meta_block are documented in the sidebar research as future enhancements but are NOT implemented. The PR description was incorrect and has been updated. When reviewing future PRs, check the actual `syncSidebar()` method — it is the source of truth for which channels are active.

**Comment by EtanHey** [comment]
@coderabbitai I need to correct a previous learning. Orphan reparenting IS needed, not just orphan survival. When a parent agent crashes (deletion_intent: crash_detected), children should have their parent_agent_id set to null (reparented to root) so they don't point to a dead agent. The previous learning that said 'do not flag or implement reparenting' was wrong — that was a V1 simplification we're now fixing. Please update the learning: orphaned children should be reparented to root (parent_agent_id = null) during the reconciliation sweep when their parent is in error state.

**Review by EtanHey — COMMENTED** [review]
## Re-evaluation: "Acceptable Deviations" Revisited

I went back through my 3-pass review and challenged every "ship as-is" / "acceptable deviation." Here's what I found:

### 1. O(n) getChildren scan → **Still correct to skip the index**

I traced every call site. `getChildren` is called in:
- `spawnAgent` (once per spawn — user-initiated, infrequent)
- `getSubtree` → `cascadeKill` (once per explicit kill — user-initiated)
- NOT in the 5s sweep hot path

Worst case at MAX_CHILDREN=10, MAX_SPAWN_DEPTH=2: 111 agents × 111 filter iterations = ~12K ops. This runs on kill, not on sweep. Adding a `Map<string, Set<string>>` children index means maintaining it across 4 code paths (set, remove, reparent, reconstitute) for zero measurable benefit at ≤111 agents. **This is genuine YAGNI.**

### 2. Kill-without-respawn for depth>0 → **Still correct, but for WRONG reasons**

My original review said "the design says so." That's lazy reasoning. The real reasons auto-respawn is wrong:

1. **Respawn loses work-in-progress context.** A depth-1 agent at 80% context has been working for a while. Killing it and spawning a fresh agent with the same prompt means the new one re-reads files, redoes analysis, potentially conflicts with uncommitted changes from the killed agent.

2. **The parent orchestrator should decide retry strategy.** Auto-respawn bypasses the parent's judgment about whether the task is done, partially done, or stuck. The parent has context the sweep doesn't.

3. **Repeated respawns accumulate dead children.** `getChildren` counts ALL children including dead ones. Spawn→80%→kill→respawn cycles would fill MAX_CHILDREN=10 with 9 corpses and 1 working agent. The parent can't spawn other workers.

Updated comments in `agent-engine.ts:178` and `quality-tracking.test.ts:261` with this reasoning (commit 86b9bde).

### 3. updateRecord not logging → **Already fixed (was correct to fix)**

Fixed in b2b4da0. Quality degradation events now appear in events.jsonl with `source: "updateRecord"`. This was the right call — silent mutations to agent state would make incident debugging impossible.

**Comment by EtanHey** [comment]
@coderabbitai review - 3 new commits since your last review: cycle detection, orphan reparenting, audit logging. Please re-review.

**Comment by EtanHey** [comment]
@coderabbitai full review

**Comment by EtanHey** [comment]
@coderabbitai review
